src/controllers/apps/chat-app/message.controllers.js

```diff
@@ -154,5 +158,89 @@ const sendMessage = asyncHandler(async (req, res) => {
     .status(201)
     .json(new ApiResponse(201, receivedMessage, "Message saved successfully"));
 });
+const deleteMessage = asyncHandler(async (req, res) => {
+  //controller to delete chat messages and attachments
 
-export { getAllMessages, sendMessage };
+  const { chatId, messageId } = req.params;
+
+  //Find the chat based on chatId
+
+  const chat = await Chat.findOne({
+    _id: new mongoose.Types.ObjectId(chatId),
+  });
+
+  if (!chat) {
+    throw new ApiError(404, "Chat does not exist");
+  }
+  //Find the message based on message id
+
+  const message = await ChatMessage.findOne({
+    _id: new mongoose.Types.ObjectId(messageId),
+  });
+
+  if (!message) {
+    throw new ApiError(404, "Message does not exist");
+  }
+
+  // Check if user is the sender of the message
+  if (message.sender.toString() !== req.user._id.toString()) {
+    throw new ApiError(
+      401,
+      "You are not authorised to delete the message ,you are not the sender"
+    );
+  }
+
+  if (
+    !chat.participants.some(
+      (user) => user?.toString() === req.user?._id.toString()
+    )
+  ) {
+    throw new ApiError(
+      401,
+      "You are not part of this group you cannot delete this message"
+    );
+  }
+  if (message.attachments.length > 0) {
+    //If the message is attachment  remove the attachments from the server
+    message.attachments.map((asset) => {
+      removeLocalFile(asset.localPath);
+    });
+  }
+  //deleting the message from DB
+  await ChatMessage.deleteOne({
+    _id: new mongoose.Types.ObjectId(messageId),
+  });
+
+  //Updating the last message of the chat to the previous message after deletion if the message deleted was last message
+  if (chat.lastMessage.toString() === message._id.toString()) {
+    const lastMessage = await ChatMessage.findOne(
+      { chat: chatId },
+      {},
+      { sort: { createdAt: -1 } }
+    );
+
+    await Chat.findByIdAndUpdate(chatId, {
+      lastMessage: lastMessage ? lastMessage?._id : null,
+    });
+  }
+  // logic to emit socket event about the message deleted  to the other participants
+  chat.participants.forEach((participantObjectId) => {
+    // here the chat is the raw instance of the chat in which participants is the array of object ids of users
+    // avoid emitting event to the user who is deleting the message
+    if (participantObjectId.toString() === req.user._id.toString()) return;
+
+    // emit the delete message event to the other participants frontend with delete messageId as the payload
+    emitSocketEvent(
+      req,
+      participantObjectId.toString(),
+      ChatEventEnum.MESSAGE_DELETED_EVENT,
+      message._id
+    );
+  });
+
+  return res
+    .status(200)
+    .json(new ApiResponse(200, {}, "Message deleted succesfully"));
```

Comment:
send message id in the response data to allow frontend to remove message from the panel

Reply:
I have added a new commit based on the review